source/dialogs/bulkconverterdialog.cpp

```diff
@@ -0,0 +1,221 @@
+
+            std::optional<std::string> error = convertFile(entry, toPath, fileFormatManager);
+            if (error != std::nullopt) {
+                QString err = QString::fromStdString(error.value());
+                emit errorMessage("error processing file: " + err);
+            }
+
+            emit progress((int) myFileCount);
+        }
+    } while (!children.empty());
+
+    emit done();
+}
+
+BulkConverterDialog::BulkConverterDialog(QWidget *parent)
+  : QDialog(parent), ui(new Ui::BulkConverterDialog),
+    myBulkWorkerThread(nullptr), myBulkConverterWorker(nullptr)
+{
+    ui->setupUi(this);
+
+    auto flags = windowFlags();
+    setWindowFlags(flags | Qt::WindowStaysOnTopHint);
+
+    connect(ui->fileLocatorSource, &QAbstractButton::clicked, [=]() {
+        setPath(ui->sourcePathEdit);
+    });
+
+    connect(ui->fileLocatorDestination, &QAbstractButton::clicked, [=]() {
+        setPath(ui->destinationPathEdit);
+    });
+
+    connect(ui->convertButton, &QAbstractButton::clicked, [=](){ convert(); });
+
+    connect(ui->exitButton, &QAbstractButton::clicked, this, &QDialog::reject);
+}
+
+BulkConverterDialog::~BulkConverterDialog()
+{
+    if (myBulkWorkerThread) {
+        myBulkWorkerThread->quit();
+        myBulkWorkerThread->wait();
+    }
+    delete ui;
+}
+
+void BulkConverterDialog::setPath(QLineEdit *edit)
+{
+    auto dialogOptions = QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks;
+    QString path = QFileDialog::getExistingDirectory(
+        this, tr("Directory To Convert"), QDir::currentPath(), dialogOptions);
+
+    edit->setText(path);
+    QString src = ui->sourcePathEdit->text();
+    QString dst = ui->destinationPathEdit->text();
+    const bool bothEditsHaveValues = !src.isEmpty() && !dst.isEmpty();
+    ui->convertButton->setEnabled(bothEditsHaveValues);
+}
+
+void BulkConverterDialog::convert()
+{
+    ui->convertButton->setEnabled(false);
+
+    boost::filesystem::path src = Paths::fromQString(ui->sourcePathEdit->text());
+    boost::filesystem::path dst = Paths::fromQString(ui->destinationPathEdit->text());
+
+    {
+        // set default max.
+        BulkConverterWorker bcw(src, dst, true);
+        bcw.walkAndConvert();
+        const std::size_t fileCountToConvert = bcw.fileCount();
+        ui->progressBar->setMaximum((int) fileCountToConvert);
+        ui->progressBar->setValue(0);
+    }
+
+    // create thread to offload work in the background, such that we can
+    // update the progress bar with each step informing the user of the
+    // current state of bulk conversion.
+    myBulkWorkerThread = new QThread;
+    myBulkConverterWorker = new BulkConverterWorker(src, dst, false);
+    myBulkConverterWorker->moveToThread(myBulkWorkerThread);
+
+    connect(myBulkWorkerThread, &QThread::started,
+            myBulkConverterWorker, &BulkConverterWorker::walkAndConvert);
+
+    connect(myBulkWorkerThread, SIGNAL(finished()),
+            myBulkWorkerThread, SLOT(deleteLater()));
```

Comment:
Just a minor note, but I like using the new-style connections (e.g. `&QObject::deleteLater` instead of `SLOT(deleteLater())` since there can be better compile-time checking

Reply:
My bad, I'm kind of learning along the way and I think I picked this out from some outdated doc! Fixed!